WordPress/gutenberg
# Adds tabs block behind block experiments flag

## What?

Adds an experimental tabs block. To enable it, turn on Experimental blocks from Gutenberg > Experiments in wp-admin.

## Why?

To enable tabbed content for sites and themes. See https://github.com/WordPress/gutenberg/issues/34079. Part of https://github.com/WordPress/gutenberg/issues/63501.

## How?

Includes 2 blocks:
 
1. core/tabs is the container, it handles rendering the tab labels. It can only contain core/tab blocks.
2. core/tab holds the content for each tab used for the tab panels. It can contain any kind of block normally used in the editor.

This first version is behind the block experiments so that we can iterate as needed without writing block deprecations.

## Testing Instructions

- Go to `/wp-admin/admin.php?page=gutenberg-experiments` and turn on Experimental blocks
- Insert a tabs block in the editor
- Change the title of the tabs and add content
- Add additional tabs
- Save and view the front end to see the tabbed content
- Check that markup contains the proper ARIA attributes described in the [Tabs Pattern guide](https://www.w3.org/WAI/ARIA/apg/patterns/tabs/#wai-ariaroles,states,andproperties)

Other features to test

- Anchor property for Tab blocks can be used to set the id for each tab (select an individual tab block in the list view then go to the advanced settings in the right sidebar)
- Selecting an inactive tab or its inner blocks from the list view will make that tab to be active
- Individual tab blocks can be dragged from one tabs block to another
- Front-end markup with JS disabled is a list of on page links

### Testing Instructions for Keyboard

In the front-end:

- Use <kbd>Tab</kbd> to select the active tab label and tab panel
- With a tab label selected, use <kbd>right arrow</kbd> and <kbd>left arrow</kbd> to select and active tabs

## Screenshots or screencast 

### Editor

https://github.com/user-attachments/assets/4559c3ed-53a6-460d-a2b3-5c112626f064

### Front-end

https://github.com/user-attachments/assets/4e919679-8066-4aa3-8fbf-c583f071d6c0

## Further work

This initial version is intentionally minimal to try and reduce the amount of code to review. Items for further work in future PRs include

- Improving inner blocks so block selection is more intuitive (right now selecting a tab label selects the tabs wrapper block, rather than the individual tab)
- When adding a new tab in the editor, focus the new tab's label
- Adding default focus styles for active tab, in case the theme doesn't have them
- Style supports for the tabs container, tab labels, and tab panel content
- Better integration anchor supports, to ensure there are no duplicate ids on the page
- A block variation for vertical tabs
- Deep linking for tab panels to open the page 
- Conditionally adding `tabindex="0"` to the active tab panel only if it does not contain content where the first element in the panel is focusable (see [ARIA example](https://www.w3.org/WAI/ARIA/apg/patterns/tabs/examples/tabs-automatic/#accessibilityfeatures))
- Adding automated unit, integration, and end-to-end tests
- Option to select the default tab that's open
- Responsive styling that stacks tabs for small viewports
